xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/PVRClient.h

```diff
@@ -0,0 +1,4763 @@
+    supportsChannelGroups = capabilities->bSupportsChannelGroups;
+    supportsChannelScan = capabilities->bSupportsChannelScan;
+    supportsChannelSettings = capabilities->bSupportsChannelSettings;
+    handlesInputStream = capabilities->bHandlesInputStream;
+    handlesDemuxing = capabilities->bHandlesDemuxing;
+    supportsRecordingPlayCount = capabilities->bSupportsRecordingPlayCount;
+    supportsLastPlayedPosition = capabilities->bSupportsLastPlayedPosition;
+    supportsRecordingEdl = capabilities->bSupportsRecordingEdl;
+    supportsRecordingsRename = capabilities->bSupportsRecordingsRename;
+    supportsRecordingsLifetimeChange = capabilities->bSupportsRecordingsLifetimeChange;
+    supportsDescrambleInfo = capabilities->bSupportsDescrambleInfo;
+    supportsAsyncEPGTransfer = capabilities->bSupportsAsyncEPGTransfer;
+
+    for (unsigned int i = 0; i < capabilities->iRecordingsLifetimesSize; ++i)
+      recordingsLifetimeValues.push_back(std::pair<int, std::string>(capabilities->recordingsLifetimeValues[i].iValue, capabilities->recordingsLifetimeValues[i].strDescription));
+  }
+
+  void setC(PVR_ADDON_CAPABILITIES& capabilities)
+  {
+    capabilities.bSupportsEPG = supportsEPG;
+    capabilities.bSupportsEPGEdl = supportsEPGEdl;
+    capabilities.bSupportsTV = supportsTV;
+    capabilities.bSupportsRadio = supportsRadio;
+    capabilities.bSupportsRecordings = supportsRecordings;
+    capabilities.bSupportsRecordingsUndelete = supportsRecordingsUndelete;
+    capabilities.bSupportsTimers = supportsTimers;
+    capabilities.bSupportsChannelGroups = supportsChannelGroups;
+    capabilities.bSupportsChannelScan = supportsChannelScan;
+    capabilities.bSupportsChannelSettings = supportsChannelSettings;
+    capabilities.bHandlesInputStream = handlesInputStream;
+    capabilities.bHandlesDemuxing = handlesDemuxing;
+    capabilities.bSupportsRecordingPlayCount = supportsRecordingPlayCount;
+    capabilities.bSupportsLastPlayedPosition = supportsLastPlayedPosition;
+    capabilities.bSupportsRecordingEdl = supportsRecordingEdl;
+    capabilities.bSupportsRecordingsRename = supportsRecordingsRename;
+    capabilities.bSupportsRecordingsLifetimeChange = supportsRecordingsLifetimeChange;
+    capabilities.bSupportsDescrambleInfo = supportsDescrambleInfo;
+    capabilities.bSupportsAsyncEPGTransfer = supportsAsyncEPGTransfer;
+
+    capabilities.iRecordingsLifetimesSize = recordingsLifetimeValues.size();
+    for (unsigned int i = 0; i < capabilities.iRecordingsLifetimesSize && i < sizeof(capabilities.recordingsLifetimeValues); ++i)
+    {
+      capabilities.recordingsLifetimeValues[i].iValue = recordingsLifetimeValues[i].first;
+      strncpy(capabilities.recordingsLifetimeValues[i].strDescription, recordingsLifetimeValues[i].second.c_str(), sizeof(capabilities.recordingsLifetimeValues[i].strDescription) - 1);
+    }
+  }
+  /*! \endcond */
+
+  /// @brief **true** if the add-on provides EPG information
+  bool supportsEPG = {false};
```

Comment:
What I do not like is that each and every member of the C-API entity is copied in its C++ counterpart (except array<->std::string as far as I can see). Wouldn't it work to derive the C++ counterpart class from the C struct? This would greatly remove complexity and increase maintainability.

For array<->std:string you could add respective methods to the C++ class, which for example converts the array of the C-API type to a std::string and vice versa.

Additionally, if an `AddonPVRCapabilities` instance actually would be a `PVR_ADDON_CAPABILITIES` there would be no need for a conversion method like `setC`.

Reply:
About the `PVR_ADDON_CAPABILITIES` / `AddonPVRCapabilities` was mainly with the `PVR_ATTRIBUTE_INT_VALUE recordingsLifetimeValues[PVR_ADDON_ATTRIBUTE_VALUES_ARRAY_SIZE];` inside, I can separate them to a own function then can be the rest stay in "C" only structure.

Reply:
With structures have e.g. `char strChannelName[PVR_ADDON_NAME_STRING_LENGTH];` or the `PVR_TIMER_TYPE`becomes it harter to make it easy to C++ without the separate C++ Class.

Because these are given by Kodi itself, it is not so easy to extend them itself.

Reply:
Not sure you understood what I mean.

Reply:
I will try to come up with a piece of sample code.

Reply:
@AlwinEsch Here is an example how I would model the C vs C++ data types:

Types definition:

```
extern "C"
{

// C-API
typedef struct PVR_ADDON_CAPABILITIES
{
  bool bSupportsEPG;

  unsigned int iRecordingsLifetimesSize;

  PVR_ATTRIBUTE_INT_VALUE recordingsLifetimeValues[PVR_ADDON_ATTRIBUTE_VALUES_ARRAY_SIZE];

} ATTRIBUTE_PACKED PVR_ADDON_CAPABILITIES;

} /* extern "C" */

namespace XXX
{

// convenience wrapper for PVR_ADDON_CAPABILITIES
class AddonPVRCapabilities : public PVR_ADDON_CAPABILITIES
{
public:
  AddonPVRCapabilities() = delete;
  AddonPVRCapabilities(const PVR_ADDON_CAPABILITIES& capabilities)
  : PVR_ADDON_CAPABILITIES(capabilities) // compiler generated copy ctor
  {}
  virtual ~AddonPVRCapabilities() = default;

  // array<->std::vector conversion convenience functions

  std::vector<std::pair<int, std::string>> GetRecordingsLifetimeValues() const
  {
    // Note: as we cannot control write access to the base class members
    //       we cannot cache convenience values in instance members of this class

    std::vector<std::pair<int, std::string>> values;
    for (unsigned int i = 0; i < iRecordingsLifetimesSize; ++i)
    {
      const int iValue = recordingsLifetimeValues[i].iValue;
      const std::string strDescr = recordingsLifetimeValues[i].strDescription;
      values.push_back({iValue, strDescr});
    }
    return values;
  }

  void SetRecordingsLifetimeValues(const std::vector<std::pair<int, std::string>>& values)
  {
    // update base struct data
    for (size_t i = 0; i < values.size(); ++i)
    {
      recordingsLifetimeValues[i].iValue = values[i].first;
      strncpy(recordingsLifetimeValues[i].strDescription,
              values[i].second.c_str(),
              sizeof(recordingsLifetimeValues[i].strDescription) - 1);
    }
  }
};

} // namespace XXX
```

Usage:

```
{
  // Construct C++ type with C type
  PVR_ADDON_CAPABILITIES caps({false, 1, {{42, "recordinglifetime1"}}});
  XXX::AddonPVRCapabilities capabilities(caps);

  // read base struct member
  bool bSupportsEPG = capabilities.bSupportsEPG;

  // write base struct member
  capabilities.bSupportsEPG = !bSupportsEPG;

  // read recordings life time values convenience data
  const std::vector<std::pair<int, std::string>> values = capabilities.GetRecordingsLifetimeValues();

  // write recordings life time values convenience data
  capabilities.SetRecordingsLifetimeValues({{42, "recordinglifetime1"}, {4711, "recordinglifetime2"}});

  // Convert C++ type to C type
  PVR_ADDON_CAPABILITIES caps2 = capabilities;
}
```
As you can see, with this approach there is no need for having two structs/classes carrying basically the same data. If you add a simple type to the C struct you need not to touch the C++ type until you want to write data conversion functions (char array <-> std::string or the like).

What do you think?

Reply:
Thanks, sound good and I change to them.

Reply:
Just a thought:
If we use inheritance we will run into problems when we want to add values to `PVR_ADDON_CAPABILITIES` without breaking binary compatibility, as per: https://community.kde.org/Policies/Binary_Compatibility_Issues_With_C%2B%2B

> You cannot ... For non-static members: add new, data members to an existing class.

It also states as thumb of rule (not sure if this still holds true):
> make all constructors non-inline.

How about following implementation:
```
extern "C"
{

// C-API
typedef struct PVR_ADDON_CAPABILITIES
{
  bool bSupportsEPG;

  unsigned int iRecordingsLifetimesSize;

  PVR_ATTRIBUTE_INT_VALUE recordingsLifetimeValues[PVR_ADDON_ATTRIBUTE_VALUES_ARRAY_SIZE];

} ATTRIBUTE_PACKED PVR_ADDON_CAPABILITIES;

} /* extern "C" */

namespace XXX
{

// convenience wrapper for PVR_ADDON_CAPABILITIES
class AddonPVRCapabilities
{
private:
  std::unique_ptr<PVR_ADDON_CAPABILITIES> m_pimpl;
public:
  AddonPVRCapabilities();
  AddonPVRCapabilities(const PVR_ADDON_CAPABILITIES& capabilities);
  virtual ~AddonPVRCapabilities();

  bool GetSupportsEPG() { return m_pimpl->bSupportsEPG; }
  void SetSupportsEPG(bool supportsEPG) { m_pimpl->bSupportsEPG = supportsEPG; }

  // TODO implement the rest
};

AddonPVRCapabilities::AddonPVRCapabilities() : m_pimpl(new PVR_ADDON_CAPABILITIES) {}

AddonPVRCapabilities(const PVR_ADDON_CAPABILITIES& capabilities)
{
  // copy construction
}

virtual ~AddonPVRCapabilities() = default; // compiler generated non-inline


} // namespace XXX
```
With this sort of implementation, we can add members to the end of the C-Structure like we did in the past and add Getters/Setters to the C++ convenient class without breaking binary compatibility.

What do you think?
Or is this a non-issue anyway?

Reply:
@a1rwulf you are completely right and i think we do not want to loose the ability to extend structs without breaking ABI. Your implementation looks correct to me. 

ctor does not need to be non-inline, though as it heap-allocates the impl struct. heap-alloc is key here, as pointer size, thus the size of the wrapper instance does not change when changing the wrapped struct.

Big plus of this approach is strict data encapsulation. C++ impl can even hide unwanted inconvenient stuff like arrays and only provide std::xxx based access methods.

Little downside is (the price of the upside just mentioned) that you have to write setters/getters for every data member you want to expose. As I wrote earlier this is IMO worth the effort.

So, I'm all in for the approach you suggested @a1rwulf.

Reply:
I'm not sure if we can use unique_ptr when everything is implemented in the header, see:
https://www.fluentcpp.com/2017/09/22/make-pimpl-using-unique_ptr
But if it's a problem, we can still use a plain pointer.

Reply:
Thought about that as well. I think using unique_ptr here should be generally okay. Just to be 100% on the safe side we should use a raw pointer here.

Reply:
I guess unique_ptr implementation is not 100% inline. So, using it in the addon header adds a runtime dependency to the c++ std library (shared library/dll) to the addon shared library which most probably is needed anyway due to usage of other types from the c++ std lib.